runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c

```diff
@@ -0,0 +1,249 @@
+  return deviceId;
+});
+
+#else
+
+WGPUAdapter wgpuInstanceRequestAdapterSync() {
+  fprintf(stderr, "wgpuInstanceRequestAdapterSync requires -sASYNCIFY\n");
+  return NULL;
+}
+
+WGPUDevice wgpuAdapterRequestDeviceSync(WGPUAdapter adapterId) {
+  fprintf(stderr, "wgpuAdapterRequestDeviceSync requires -sASYNCIFY\n");
+  return NULL;
+}
+
+#endif  // EM_ASYNC_JS
+
+//===----------------------------------------------------------------------===//
+// iree_hal_webgpu_emscripten_driver_t
+//===----------------------------------------------------------------------===//
+
+typedef struct iree_hal_webgpu_emscripten_driver_t {
+  iree_hal_resource_t resource;
+  iree_allocator_t host_allocator;
+
+  iree_string_view_t identifier;
+  iree_hal_webgpu_device_options_t default_options;
+
+  WGPUInstance instance;
+  WGPUAdapter adapter;
+} iree_hal_webgpu_emscripten_driver_t;
+
+static const iree_hal_driver_vtable_t iree_hal_webgpu_emscripten_driver_vtable;
+
+static iree_hal_webgpu_emscripten_driver_t*
+iree_hal_webgpu_emscripten_driver_cast(iree_hal_driver_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_webgpu_emscripten_driver_vtable);
+  return (iree_hal_webgpu_emscripten_driver_t*)base_value;
+}
+
+iree_status_t iree_hal_webgpu_emscripten_driver_create(
+    iree_string_view_t identifier,
+    const iree_hal_webgpu_driver_options_t* options,
+    iree_allocator_t host_allocator, iree_hal_driver_t** out_driver) {
+  IREE_ASSERT_ARGUMENT(options);
+  IREE_ASSERT_ARGUMENT(out_driver);
+  *out_driver = NULL;
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_webgpu_emscripten_driver_t* driver = NULL;
+  iree_host_size_t total_size = sizeof(*driver) + identifier.size + /*NUL=*/1;
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_allocator_malloc(host_allocator, total_size, (void**)&driver));
+  iree_hal_resource_initialize(&iree_hal_webgpu_emscripten_driver_vtable,
+                               &driver->resource);
+  driver->host_allocator = host_allocator;
+
+  iree_string_view_append_to_buffer(identifier, &driver->identifier,
+                                    (char*)driver + sizeof(*driver));
+  memcpy(&driver->default_options, &options->device_options,
+         sizeof(driver->default_options));
+
+  const WGPUInstanceDescriptor instance_descriptor = {
+      .nextInChain = NULL,
+  };
+  driver->instance = wgpuCreateInstance(&instance_descriptor);
+  if (!driver->instance) {
+    iree_hal_driver_release((iree_hal_driver_t*)driver);
+    IREE_TRACE_ZONE_END(z0);
+    return iree_make_status(
+        IREE_STATUS_UNAVAILABLE,
+        "WebGPU implementation not present or failed to load");
+  }
+
+  // Request an adapter from the implementation. We only get one of these and it
+  // may expose multiple devices so it's effectively what we consider a driver.
+  // HACKS: sync via Asyncify, assume success
+  WGPUAdapter adapter = wgpuInstanceRequestAdapterSync();
+  IREE_ASSERT_NE(adapter, NULL);
```

Comment:
worth adding a check here instead - having webgpu doesn't meant you have an adapter, AFAIK (though maybe browsers hide webgpu entirely if no adapters would be returned - seems like that'd be expensive though)